packages/block-editor/src/components/link-control/README.md

```diff
@@ -1,12 +1,28 @@
 # Link Control
 
+Renders a link control. A link control is a controlled input which maintains
+a value associated with a link (HTML anchor element) and relevant settings
+for how that link is expected to behave.
+
 ## Props
 
 ### value
 
 - Type: `Object`
 - Required: No
 
+Current link value.
+
+A link value contains is composed as a union of the default properties and any custom settings values.
+
+Default properties include:
+
+- id (`string|number`): Unique identifier of link.
+- url (`string`): Link URL.
+- title (`string`): Link title.
+- type (`string`): Type of link entity.
```

Comment:
What are the possible values for type and subtype?

Reply:
> What are the possible values for type and subtype?

My understanding of the way it works today is that this inherits the same schema as the REST API search endpoint:

https://developer.wordpress.org/rest-api/reference/search-results/

(You can see there the possible values for this today)

I think a natural worry here could be that it ties too closely with the WordPress-specific idea of what a link can represent, but it seems to have been designed in a fairly generic sense, and I think this could extend to represent other "type"s of link entities.

#19827 is similar here in trying to group all manually-entered URLs under a "url" type.

Also some related discussion at https://github.com/WordPress/gutenberg/pull/19387#issuecomment-576765555.

Reply:
It might help to document _something_ here about how someone should expect to use these, but I'm not entirely clear what exactly. It should probably tie into the introduction of documentation surrounding the `fetchLinkSuggestions` context.

Reply:
Is there a "type" for raw url?

Reply:
As documented here, there would have to be something yes, even a subtype must be assigned.

Again, this is inherited largely from the [schema for the search endpoint](https://developer.wordpress.org/rest-api/reference/search-results/).

I might be inclined to think that it probably makes sense to allow them to be nullable or omitted, but we'd want to be quite explicit about that, and I think it would require one of:

- Diverging from the search endpoint schema
- Updating the search endpoint schema

Ties in as well with:

- #19827
  - >I have doubts about whether we really care to assign these subtypes, vs. leaving it to the consumer to deal with a URL as they see fit
- #19651
  - I leave out all properties other than `url` in the raw, submitted URL in e6083793ec50fb6b7cd93d9e08cdce7ce5de1168
  - This introduces a bug in how those suggestions are merged to the current value, fixed in #19827. I think it's symptomatic of issues that can arise when there are inconsistencies in the value shape.

Reply:
Nothing blocking here feel free to move forward but yeah the clearer the value format is, the easier it is to work with the component.

Reply:
I'm glad you're thinking about it, and I do think it's an important decision to make, considering its impact on usability and consistency. I'm not sure how common it would be to manually create values for the component, but I don't like as in the second code snippet of https://github.com/WordPress/gutenberg/issues/18061#issuecomment-577334493 that it would be considered invalid when providing a value to not also include the `type`, `subtype`, or even `title` of a link. I'd like to think it should be perfectly acceptable to pass `{ url: 'http://example.com' }` as a value.

This might be something where a selected _suggestion_ would be expected include these properties, but that it's not necessarily required for the value provided. It's a bit confusing, especially if we consider that most usage might look something like:

```js
const [ value, setValue ] = useState();
return <LinkControl value={ value } onChange={ ( nextValue ) => setValue( nextValue ) } />;
```

With this code, we'd probably consider `nextValue` to be of the same general shape as the `value` passed.

It's also a question of: What of these details does LinkControl actually use when a value is assigned, and does it require them?

- For `type`, `id`, and `subtype`, the component doesn't actually use them
- For `title`, the component uses them, but does (or should) have fallbacks when it's missing or empty
- `url` is the one which is both used and required

I wonder if the best way to move forward on this specific pull request would be to only document the `title` and `url` parts of the value, perhaps considering in a separate pull request to add a different "suggestion" type which can include the ID, type, and subtype.

`title` is an interesting one, because we probably want to allow it to be optional. But does that mean:

- Allow it to be omitted altogether
- Make it required, but allow it to be `null`
- Make it required, and it must be a string (use an empty string `''` if it's not assigned)

The first is probably the easiest from usability, but further reinforces this difference between a value type and a suggestion type.

Maybe, in the end, the best compromise would be one where the value type and suggestion type are technically distinct, but that `value` is aware of and would optionally accept any properties from a suggestion, even if it didn't strictly need or use them.

Example: https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgOoAUAyoDWBhAe3CgIBsA1OUgVxQG8BYAKGVeWABMB+ALmQGcwUUAHNkAH2QhqAWwBG0ADTM2yalFJ9BwkCOUs2YYGFIReAoaP2qwATwAOZrZd3W2-anLuPz2q8wBfZmZQSFhEFAxsEHwiITIAZWoREQhBYCJkCAAPSBAOfjQsXEJiMkoaehU2TmcdMUlpeSVq1iMTCDr-AzaHTot6t1YPLz6u10DgpgB6aeQAWkWl5ZWphCJBZAA3Klo+KJK4kgpdlABeZDo1DT4AcgALMDB7HlmcuBl7UwA6dZlb5ABADczHWIE2IByYAqeyK0ViZVISRSaSMmQujB67A4d0ez1e03enx+f1uQ2ummQDyeLze2Q+XwgvwI-3J7VMdzJrWQ3n6t3UpC5WJGvNxNKFwLWGzA21OqGM9wA8vY0SAqABRXJQOD7YoxUrxE6VZAY7m1Kl42mE+nEpmk8kCsX4ukMkksoU2YwcqkewxjKkC33DTyii3i-SSphgzagHakTgAOShMP6B31R0SyVS6XRlwpTqtRMZzP+gKBQA

I still think it's awkward to try to define `id`, `subtype` for a "raw URL" value 🤔

Reply:
I was going to reply to your comment and noticed that I think we're on the same page for everything I think.

> I wonder if the best way to move forward on this specific pull request would be to only document the title and url parts of the value, perhaps considering in a separate pull request to add a different "suggestion" type which can include the ID, type, and subtype.

While using the component, I didn't even know it contained type, suptype and id. so yeah title and url are more important.

> The first is probably the easiest from usability, but further reinforces this difference between a value type and a suggestion type.

Is that bad? I think it's ok for different "types" of links. I also considered whether we should have a `type: "raw"` to clarify the difference but I guess the absence of "type" could mean the same thing.

Reply:
> > The first is probably the easiest from usability, but further reinforces this difference between a value type and a suggestion type.
> 
> Is that bad? I think it's ok for different "types" of links. I also considered whether we should have a `type: "raw"` to clarify the difference but I guess the absence of "type" could mean the same thing.

Not necessarily, but I think it will be natural to observe that the two types share a fair few number of properties, and conclude that they are somehow related (or, the same).